main/src/ca/uwaterloo/flix/language/phase/Safety.scala

```diff
@@ -717,53 +821,54 @@ object Safety {
   /**
     * Checks for `IllegalRelationalUseOfLatticeVariable` in the given `head` predicate.
     */
-  private def checkHeadPredicate(head: Predicate.Head, latVars: Set[Symbol.VarSym]): List[CompilationMessage] = head match {
-    case Predicate.Head.Atom(_, Denotation.Latticenal, terms, _, loc) =>
+  private def checkHeadPredicate(head: Predicate.Head, latVars: Set[Symbol.VarSym]): Validation[Predicate.Head, CompilationMessage] = head match {
+    case Predicate.Head.Atom(pred, Denotation.Latticenal, terms, tpe, loc) =>
       // Check the relational terms ("the keys").
-      checkTerms(terms.dropRight(1), latVars, loc)
-    case Predicate.Head.Atom(_, Denotation.Relational, terms, _, loc) =>
+      mapN(checkTerms(terms.dropRight(1), latVars, loc))(Predicate.Head.Atom(pred, Denotation.Latticenal, _, tpe, loc))
+    case Predicate.Head.Atom(pred, Denotation.Relational, terms, tpe, loc) =>
       // Check every term.
-      checkTerms(terms, latVars, loc)
+      mapN(checkTerms(terms, latVars, loc))(Predicate.Head.Atom(pred, Denotation.Relational, _, tpe, loc))
   }
 
   /**
     * Checks that the free variables of the terms does not contain any of the variables in `latVars`.
     * If they do contain a lattice variable then a `IllegalRelationalUseOfLatticeVariable` is created.
     */
-  private def checkTerms(terms: List[Expression], latVars: Set[Symbol.VarSym], loc: SourceLocation): List[CompilationMessage] = {
+  private def checkTerms(terms: List[Expression], latVars: Set[Symbol.VarSym], loc: SourceLocation): Validation[List[Expression], CompilationMessage] = {
     // Compute the free variables in all terms.
     val allVars = terms.foldLeft(Set.empty[Symbol.VarSym])({
       case (acc, term) => acc ++ freeVars(term).keys
     })
 
     // Compute the lattice variables that are illegally used in the terms.
-    allVars.intersect(latVars).toList.map(sym => IllegalRelationalUseOfLatticeVariable(sym, loc))
+    mapN(traverse(allVars.intersect(latVars).toList)(IllegalRelationalUseOfLatticeVariable(_, loc).toFailure)) { _ => terms }
```

Comment:
I think the structure of this function will need to change in order to go over the expressions individually. You can probably leave a TODO though.

Reply:
Alright.